reaktive/src/commonTest/kotlin/com/badoo/reaktive/observable/TakeObservableTests.kt

```diff
@@ -0,0 +1,109 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.test.base.assertDisposed
+import com.badoo.reaktive.test.base.assertError
+import com.badoo.reaktive.test.base.assertNotError
+import com.badoo.reaktive.test.observable.*
+import kotlin.test.Test
+import kotlin.test.assertFailsWith
+
+class TakeObservableTests :
+    ObservableToObservableTests by ObservableToObservableTests<Unit>({ take(1) }) {
+
+    private val upstream = TestObservable<Int?>()
+    val observer = upstream.take(10).test()
+
+    @Test
+    fun completes_WHEN_limit_of_1_is_reached() {
+        val observer = upstream.take(1).test()
+
+        upstream.onNext(42)
+
+        observer.assertComplete()
+        observer.assertNotError()
+    }
+
+    @Test
+    fun completes_WHEN_limit_of_2_is_reached() {
+        val observer = upstream.take(2).test()
+
+        upstream.onNext(42)
+        observer.assertNotComplete()
+
+        upstream.onNext(84)
+        observer.assertComplete()
+        observer.assertNotError()
+    }
+
+    @Test
+    fun completes_WHEN_upstream_completed() {
+        upstream.onComplete()
+
+        observer.assertComplete()
+    }
+
+    @Test
+    fun produces_same_error_WHEN_upstream_produced_error() {
+        val error = Exception()
+
+        upstream.onError(error)
+
+        observer.assertError(error)
+    }
+
+    @Test
+    fun disposes_WHEN_upstream_disposed() {
+        upstream.dispose()
+
+        observer.assertDisposed()
+    }
+
+    @Test
+    fun produces_values_in_correct_order_from_upstream() {
+        upstream.onNext(42, null, 84, null)
+
+        observer.assertValues(42, null, 84, null)
+    }
+
+    @Test
+    fun multiple_takes() {
+        val observer = upstream
+            .take(100)
+            .take(1)
+            .test()
+
+        (1..100).forEach {
+            upstream.onNext(it)
+        }
+
+        observer.assertValue(1)
+        observer.assertComplete()
+    }
+
+    @Test
+    fun re_entrant_take() {
```

Comment:
Please specify at least the expected behavior in names. Like `counts_recursive_invocations` or similar.

Reply:
I actually couldn't think of a good name. Wanna help out? 

The usecase I'm trying to test here is when `upstream` produces a value from inside `onNext()`.

Reply:
So for this one I think 'counts_recursive_invocations' would be more meaningful.